### PR TITLE
#12044: Add vc support and tests to tunneling components

### PR DESCRIPTION
**Pull request by ubcheema**
Add VCs to support independent communication between CQs and respective tunnel devices.
4 tunnel stops X 2 CQs each requires 8 forward VCs.

### CODE REVIEW

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_vc_bi_tunnel.cpp [hunk 1]
```diff
@@ -0,0 +1,845 @@
...
+        log_info(LogTest,
+            "Tx/Rx Device {}. Tunneling Ethernet core = {}.",
+            device_id_l, tunneler_logical_core.str());
+
+        log_info(LogTest,
+            "Loopback Device {}. Tunneling Ethernet core = {}.",
+            device_id_r, r_tunneler_logical_core.str());
+
+        //std::cout<<"Left Tunneler = "<<tunneler_logical_core.str()<<std::endl;
+        //std::cout<<"Right Tunneler = "<<r_tunneler_logical_core.str()<<std::endl;
```
Comment (tt-aho) [inline tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_vc_bi_tunnel.cpp:174]: Remove

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_vc_bi_tunnel.cpp [hunk 2]
```diff
@@ -0,0 +1,845 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (tt-aho) [inline tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_vc_bi_tunnel.cpp:1]: Use 2024 for new files

File: tt_metal/impl/device/device.cpp [hunk 3]
```diff
@@ -885,20 +928,35 @@ void Device::update_workers_build_settings(std::vector<std::vector<std::tuple<tt
...
+                        compile_args[9 + (demux_output_idx + i) * 2] = us_tunneler_remote_settings.cb_size_bytes >> 4;   // 11: remote_tx_queue_size_words 1
+                    }
+                } else {
+                    TT_ASSERT(fwd_vc_count == 0, "Unhandled Forward VCs encountered.");
+                }
+
+                //compile_args[16] = tunneler_settings.worker_physical_core.x; // 16: remote_rx_x
+                //compile_args[17] = tunneler_settings.worker_physical_core.y; // 17: remote_rx_y
+                //compile_args[18] = 2; // 18: remote_rx_queue_id
+                //compile_args[19] = (uint32_t)DispatchRemoteNetworkType::NOC0; // 19: tx_network_type
```
Comment (tt-aho) [inline tt_metal/impl/device/device.cpp:951]: Remove?

File: tt_metal/impl/dispatch/kernels/packet_queue.hpp [hunk 4]
```diff
@@ -181,12 +194,12 @@ class packet_queue_state_t {
...
     }
 
-    inline void reset_queue_local_wptr() {
-        *this->local_wptr_reset = 0;
-    }
-
     inline void advance_queue_local_wptr(uint32_t num_words) {
+        if (this->queue_is_input) {
         *this->local_wptr_update = num_words << REMOTE_DEST_BUF_WORDS_FREE_INC;
+        } else {
```
Comment (tt-aho) [inline tt_metal/impl/dispatch/kernels/packet_queue.hpp:200]: Indentation

File: tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+            }
+        }
+        all_outputs_finished = true;
+        for (uint32_t i = 0; i < tunnel_lanes; i++) {
+            if (input_queues[i].get_curr_packet_valid()) {
+                bool full_packet_sent;
+                uint32_t words_sent =
+                    output_queues[i].forward_data_from_input(0, full_packet_sent, input_queues[i].get_end_of_cmd());
+                // data_words_sent += words_sent;
+                // if ((words_sent > 0) && (timeout_cycles > 0)) {
```
Comment (tt-aho) [inline tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp:247]: Remove?

**Review by tt-aho — APPROVED** [review]
I'm not very familiar with mux/demux logic so can't really give a thorough review. Didn't see anything obviously wrong though

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_vc_bi_tunnel.cpp (hunk 1 above)
Comment (ubcheema) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_vc_bi_tunnel.cpp]: Fixed.

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_vc_bi_tunnel.cpp (hunk 2 above)
Comment (ubcheema) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_vc_bi_tunnel.cpp]: Fixed

File: tt_metal/impl/device/device.cpp (hunk 3 above)
Comment (ubcheema) [reply tt_metal/impl/device/device.cpp]: Fixed.

File: tt_metal/impl/dispatch/kernels/packet_queue.hpp (hunk 4 above)
Comment (ubcheema) [reply tt_metal/impl/dispatch/kernels/packet_queue.hpp]: Fixed.

File: tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp (hunk 5 above)
Comment (ubcheema) [reply tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp]: Fixed.

File: tt_metal/impl/dispatch/kernels/vc_packet_router.cpp [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+        write_test_results(test_results, PQ_TEST_MISC_INDEX, 0xff000003);
+        for (uint32_t i = 0; i < router_lanes; i++) {
+            input_queues[i].send_remote_finished_notification();
+        }
+    }
+
+    set_64b_result(test_results, data_words_sent, PQ_TEST_WORD_CNT_INDEX);
+    set_64b_result(test_results, cycles_elapsed, PQ_TEST_CYCLES_INDEX);
+    set_64b_result(test_results, iter, PQ_TEST_ITER_INDEX);
+
```
Comment (tt-asaigal) [inline tt_metal/impl/dispatch/kernels/vc_packet_router.cpp:299]: NIT: the first 2 results we write are updated/computed on the critical path. Not high overhead, but do we need these in production mode?

File: tt_metal/impl/dispatch/kernels/packet_queue.hpp [hunk 7]
```diff
@@ -508,20 +529,22 @@ class packet_input_queue_state_t : public packet_queue_state_t {
...
         this->curr_packet_valid = false;
-        this->reset_queue_local_rptr_sent();
-        this->reset_queue_local_rptr_cleared();
         this->reset_queue_local_wptr();
         this->reset_ready_flag();
     }
 
     inline uint32_t get_end_of_cmd() const {
-        return this->end_of_cmd;
+        return (uint32_t)this->end_of_cmd;
```
Comment (tt-asaigal) [inline tt_metal/impl/dispatch/kernels/packet_queue.hpp:537]: Could we return a uint16 here and make `forward_data_from_input` accept a uint16?

**Review by tt-asaigal — APPROVED** [review]
Looks good overall. Left a few tiny comments.

File: tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+
+// careful, may be null
+tt_l1_ptr uint32_t* const test_results = reinterpret_cast<tt_l1_ptr uint32_t*>(test_results_buf_addr_arg);
+
+constexpr uint32_t timeout_cycles = get_compile_time_arg_val(46);
+constexpr uint32_t inner_stop_mux_d_bypass = get_compile_time_arg_val(47);
+
+void kernel_main() {
+    rtos_context_switch_ptr = (void (*)())RtosTable[0];
+    //noc_init();
```
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp:185]: remove this so we don't reintroduce the bug with calling this in the mux/demux path

File: tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+
+    write_test_results(test_results, PQ_TEST_STATUS_INDEX, PACKET_QUEUE_TEST_STARTED);
```
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp:187]: Not super familiar with mux/demux kernels and I know this api exists rn but is `write_test_results` a good name for what this is doing?

**Review by abhullar-tt — APPROVED** [review]
lgtm, should some of the tests you added be included in `run_cpp_fd2_tests.sh`?

File: tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp (hunk 9 above)
Comment (ubcheema) [reply tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp:187]: No, it's not a good name.
Also the kernel is not a test kernel anyway.
This needs some cleaning up.

Comment (ubcheema) [reply tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp:187]: all the packet/tunneling kernels have this common quirk.
I'll clean it up.
write_kernel_status ( ) might be a better name.

Comment (ubcheema) [reply tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp]: Fixed

File: tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp (hunk 8 above)
Comment (ubcheema) [reply tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp]: Fixed.

File: tt_metal/impl/dispatch/kernels/packet_queue.hpp (hunk 7 above)
Comment (ubcheema) [reply tt_metal/impl/dispatch/kernels/packet_queue.hpp]: Fixed.

File: tt_metal/impl/dispatch/kernels/packet_queue.hpp [hunk 10]
```diff
@@ -197,25 +210,26 @@ class packet_queue_state_t {
...
         *this->local_rptr_cleared_update = num_words << REMOTE_DEST_BUF_WORDS_FREE_INC;
+        }
     }
 
-    inline uint32_t get_queue_data_num_words_available_to_send() const {
-        return (this->get_queue_local_wptr() - this->get_queue_local_rptr_sent()) & this->queue_size_mask;
-    }
+//    inline uint32_t get_queue_data_num_words_available_to_send() const {
+//        return (this->get_queue_local_wptr() - this->get_queue_local_rptr_sent()) & this->queue_size_mask;
+//    }
```
[the remark this answers is not in the document]
Comment (ubcheema) [reply tt_metal/impl/dispatch/kernels/packet_queue.hpp]: Fixed.

File: tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp (hunk 8 above)
Comment (ubcheema) [reply tt_metal/impl/dispatch/kernels/vc_eth_tunneler.cpp]: Will add tests to run_cpp_fd2_tests.sh in a separate PR.
tunneling tests need to be skipped on N150 boards.

**Pull request merged**